// ngx-cpf-cnpj/esm2015/lib/cpf-cnpj-validator.directive.js
var CpfCnpjValidatorDirective_1;
import { __decorate, __metadata } from "tslib";
import { Directive } from "@angular/core";
import { NG_VALIDATORS } from "@angular/forms";
import { NgxCpfCnpj } from "./ngx-cpf-cnpj";
let CpfCnpjValidatorDirective = CpfCnpjValidatorDirective_1 = class CpfCnpjValidatorDirective {
    constructor() {
        this.validator = this.cpfCnpjValidator();
    }
    validate(c) {
        return this.validator(c);
    }
    cpfCnpjValidator() {
        return (c) => {
            let value = c.value;
            if (value != null) {
                value = NgxCpfCnpj.getDigitos(value);
                if (value.length === 11) {
                    if (NgxCpfCnpj.cpfIsValid(value)) {
                        return null;
                    }
                    else {
                        return {
                            cpfcnpjvalidator: {
                                valid: false
                            }
                        };
                    }
                }
                else if (value.length === 14) {
                    if (NgxCpfCnpj.cnpjIsValid(value)) {
                        return null;
                    }
                    else {
                        return {
                            cpfcnpjvalidator: {
                                valid: false
                            }
                        };
                    }
                }
            }
            return {
                cpfcnpjvalidator: {
                    valid: false
                }
            };
        };
    }
};
CpfCnpjValidatorDirective = CpfCnpjValidatorDirective_1 = __decorate([
    Directive({
        selector: "[nccCpfCnpjValidator] [ngModel]",
        providers: [
            {
                provide: NG_VALIDATORS,
                useExisting: CpfCnpjValidatorDirective_1,
                multi: true
            }
        ]
    }),
    __metadata("design:paramtypes", [])
], CpfCnpjValidatorDirective);
export { CpfCnpjValidatorDirective };
