// ngx-cpf-cnpj/esm5/lib/cpf-cnpj-validator.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive } from "@angular/core";
import { NG_VALIDATORS } from "@angular/forms";
import { NgxCpfCnpj } from "./ngx-cpf-cnpj";
var CpfCnpjValidatorDirective = /** @class */ (function () {
    function CpfCnpjValidatorDirective() {
        this.validator = this.cpfCnpjValidator();
    }
    CpfCnpjValidatorDirective_1 = CpfCnpjValidatorDirective;
    CpfCnpjValidatorDirective.prototype.validate = function (c) {
        return this.validator(c);
    };
    CpfCnpjValidatorDirective.prototype.cpfCnpjValidator = function () {
        return function (c) {
            var value = c.value;
            if (value != null) {
                value = NgxCpfCnpj.getDigitos(value);
                if (value.length === 11) {
                    if (NgxCpfCnpj.cpfIsValid(value)) {
                        return null;
                    }
                    else {
                        return {
                            cpfcnpjvalidator: {
                                valid: false
                            }
                        };
                    }
                }
                else if (value.length === 14) {
                    if (NgxCpfCnpj.cnpjIsValid(value)) {
                        return null;
                    }
                    else {
                        return {
                            cpfcnpjvalidator: {
                                valid: false
                            }
                        };
                    }
                }
            }
            return {
                cpfcnpjvalidator: {
                    valid: false
                }
            };
        };
    };
    var CpfCnpjValidatorDirective_1;
    CpfCnpjValidatorDirective = CpfCnpjValidatorDirective_1 = __decorate([
        Directive({
            selector: "[nccCpfCnpjValidator] [ngModel]",
            providers: [
                {
                    provide: NG_VALIDATORS,
                    useExisting: CpfCnpjValidatorDirective_1,
                    multi: true
                }
            ]
        }),
        __metadata("design:paramtypes", [])
    ], CpfCnpjValidatorDirective);
    return CpfCnpjValidatorDirective;
}());
export { CpfCnpjValidatorDirective };
